### ISSUE TITLE
Rolling out nullness annotations

### ISSUE THREAD

**Issue by kerams** [fable-compiler/fable-browser#140]
.NET 9 is around the corner, and with it support for consuming and emitting nullness metadata in F# projects. JS APIs will often return `null`/`undefined`, and while Fable bindings often capture this information by using options, but [sometimes we give no indication that `null` are a possibility](https://github.com/fable-compiler/fable-browser/issues/99). I think F# 9 will be really useful here, so I'd like to start a discussion about the adoption plan for these annotations.

From the technical point of view, binding projects need to be compiled using .NET 9 SDK, `<LangVersion>preview</LangVersion>` (for now) and with the `--checknulls+` flag (also for now - I think this will be aligned with C#'s `<Nullable>enable</Nullable>`). TFMs don't matter, so `netstandard2.0` assemblies will contain nullness metadata too.

Finally, every API will need to be manually reviewed and annotated as required. In the case of existing bindings that use options with reference types, there's the question of whether we want to preserve options or replace them with `null` annotations, just so that a single, unified approach is used.

**Comment by MangelMaxime** [comment]
Not only the binding needs to be updated to support nullness supports but we will also need to update the FCS fork that Fable use so we can get access to the language feature.

Are the configuration you mentioned to be set on the binding project or do the user project also needs to set it? What I mean is what happens if we decide to use nullness for the binding but the user project doesn't active the language feature on its project?

Will he still have the error generated?

**Comment by kerams** [comment]
> we will also need to update the FCS fork that Fable use so we can get access to the language feature

True, but that should be easy enough. Also, you'll get warning in the IDE regardless.

> Are the configuration you mentioned to be set on the binding project or do the user project also needs to set it?

Consumers of Fable bindings also need to enable nullness checking in their projects, meaning we don't need to publish a separate set of packages with nullness metadata - it's business as usual for users that do not opt into the feature, even with assemblies enriched with nullness annotations.

**Comment by MangelMaxime** [comment]
I am tracking the nullness supports in Glutinum project [here](https://github.com/glutinum-org/cli/issues/121). It ask questions regarding how things should be represented.

It could also be a good time for checking how Glutinum handle Dom APIs. I always had the vision to be able to generate such APIs because they are kind of big and being able to generate them means we should be able to applied optimisation to them over time.

The original version of Fable.Browser, was actually generated and over time we stopped doing that because the project was splitted in different package. It was to simply the maintenance I believe but over time we discovered that it was not so much the case.

**Comment by kerams** [comment]
If you switch an existing option annotation to `| null`, you'll get a compile-time breaking change, so no silent regressions.

**Comment by MangelMaxime** [comment]
Ah yes right, this is one of the reason we use F# after all 😅 (sometimes you forget some of the simplest things).

**Comment by kerams** [comment]
> Ah, I just realized Fable packages include source files, this makes things much more complicated.

Since we have to distribute source files, I think the only option is to use conditional compilation for annotations. Something like

```fsharp
#if NET9_OR_GREATER
abstract getElementById: elementId: string | null -> HTMLElement
#else
abstract getElementById: elementId: string -> HTMLElement
#endif
```

but then we'll also need to add `net9.0` TFMs so that nullness warnings show up in the IDE, and users also need to retarget their Fable projects to `net9.0`.

**Comment by MangelMaxime** [comment]
If we need to duplicate the API surface to support nullness then I will probably vote against it, we need to either commit to it or keep using Option for the time being.

Duplicating the API surface means depending on your configuration you have a different API which his confusing
and documentation become harder to write.

I wonder why for this new feature people need to use target `net9.0`? I mean in previous addition to F# people where able to keep using `netstandard2.1` for example without issues.

If I remember how TFM works, in order to use net9.0 in your project you need to use .NET 9.0 SDK. And this is problematic because it means forcing people to upgrade to .NET 9.0 which is not a LTS version. I think people tend to avoid upgrading in general and favor LTS. 

That NET 9.0 upgrade is not so much a problem for Fable because we don't really respect/use that information. But it is a problem for Server code I think.

------------

I suppose the fact that F# distribute source files, is what prevents this part of the RFC to works:

> Older F# components consuming F# assemblies that do have nullability
> 
> F# components are no different than non-F# components when it comes to being consumed by an older F# codebase. The nullability attribute is simply ignored by older F# compilers.

[RFC Source](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1060-nullable-reference-types.md#older-f-components-consuming-f-assemblies-that-do-have-nullability)

------------

Depending what can or cannot be done, I think it is also possible to use a custom type to avoid duplicating the code:

```fs
type Nullable<'T> =
    #if NET9_0_OR_GREATER
    'T | null
    #else
    'T option
    #endif
```

It still provides 2 different API depending on the configuration but at least doesn't duplicate code in the binding/library code.
